File: phoenix/lib/ash/repositories/category_repository.ex
defmodule Ash.Repositories.CategoryRepository do
  alias Ash.Models.Category

  @categories [
    Category.new("1", "Dairy"),
    Category.new("2", "Fruit"),
    Category.new("3", "Vegetables"),
    Category.new("4", "Bakery"),
    Category.new("5", "Meat")
  ]

  def get_all_categories do
    @categories
  end

  def get_category_by_id(id) do
    Enum.find(@categories, fn category -> category.id == id end)
  end
end

File: gin/repositories/category_repository.go
package repositories

import "gin-benchmark/models"

type CategoryRepository struct {
	categories []*models.Category
}

func NewCategoryRepository() *CategoryRepository {
	return &CategoryRepository{
		categories: []*models.Category{
			{Id: "1", Name: "Dairy"},
			{Id: "2", Name: "Fruit"},
			{Id: "3", Name: "Vegetables"},
			{Id: "4", Name: "Bakery"},
			{Id: "5", Name: "Meat"},
		},
	}
}

func (r *CategoryRepository) GetAllCategories() []*models.Category {
	return r.categories
}

func (r *CategoryRepository) GetCategoryById(id string) *models.Category {
	for _, category := range r.categories {
		if category.Id == id {
			return category
		}
	}
	return nil
}
